### PR TITLE
[Feature] Hours This Week

### PR DESCRIPTION
**Pull request by whoiswillma**
## Overview

Add hours this week, which allows users to see hours for a dining hall over the next seven days. 

## Changes Made

I created the `HoursThisWeek` view that displays hours this week. 

I also added a protocol to represent views that change their content size, since there are now two of them.

I also did an `npm install`, and that seems to have forced the package-lock.json to change. If anyone knows how to lock npm packages please let me know.

## Test Coverage

Ran on simulator with various sized devices and on a physical device.

## Screenshots

<details>

  <summary>Videa</summary>

cropped mp4

  
  

</details>

### CODE REVIEW

File: Eatery/Views/Eateries/CampusMenuInfoView.swift [hunk 1]
```diff
@@ -73,13 +87,47 @@ class CampusMenuInfoView: UIView, MenuInfoView {
...
+        }
     }
 
     required init?(coder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
 
-    func configure(eatery: Eatery, userLocation: CLLocation?) {
+    override func didMoveToWindow() {
+        // defer layout until view has been added to view hierarchy
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/CampusMenuInfoView.swift]: `// Defer layout ...`

File: Eatery/Views/Eateries/CampusMenuInfoView.swift [hunk 2]
```diff
@@ -73,13 +87,47 @@ class CampusMenuInfoView: UIView, MenuInfoView {
...
+        }
+
+        locationLabel.isOpaque = false
+        locationLabel.font = UIFont.systemFont(ofSize: 14, weight: .medium)
+        locationLabel.textColor = .gray
+        addSubview(locationLabel)
+        locationLabel.snp.makeConstraints { make in
+            make.leading.equalToSuperview().offset(16)
+            make.top.equalTo(hoursThisWeekContainer.snp.bottom).offset(8)
+            make.bottom.equalToSuperview().inset(16)
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/CampusMenuInfoView.swift]: i believe you can combine `make.leading.bottom.equalToSuperview().inset(16)`

File: Eatery/Views/Eateries/HoursThisWeekView.swift [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private func setUpConstraints() {
+        segmentedControl.snp.makeConstraints { make in
+            make.leading.trailing.equalToSuperview().inset(16)
+            make.top.equalToSuperview().inset(12)
+        }
+
+        segmentedControl.snp.prepareConstraints { make in
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/HoursThisWeekView.swift]: may be better to rename `make` to `prepare`

File: Eatery/Views/Eateries/HoursThisWeekView.swift [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        }
+
+        let contentLayoutGuide = UILayoutGuide()
+        addLayoutGuide(contentLayoutGuide)
+        contentLayoutGuide.snp.makeConstraints { make in
+            make.leading.trailing.equalToSuperview().inset(16)
+            make.bottom.equalToSuperview().inset(8)
+        }
+
+        contentLayoutGuide.snp.prepareConstraints { make in
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/HoursThisWeekView.swift]: same here

**Review by kevinchan159 — APPROVED** [review]
this looks amazing!!! left some very minor comments

File: Eatery/Controllers/Eateries/Collegetown/CollegetownMenuViewController.swift [hunk 5]
```diff
@@ -40,6 +40,21 @@ class CollegetownMenuViewController: EateriesMenuViewController {
         addMapView()
     }
 
+    private func addMenuInfoView() {
+        let infoView = CollegetownMenuInfoView()
+        infoView.configure(eatery: eatery, userLocation: userLocation)
+
```
Comment (danielVebman) [inline Eatery/Controllers/Eateries/Collegetown/CollegetownMenuViewController.swift]: Remove newline?

File: Eatery/Views/Eateries/CampusMenuInfoView.swift [hunk 6]
```diff
@@ -54,14 +67,15 @@ class CampusMenuInfoView: UIView, MenuInfoView {
...
 
-        locationLabel.isOpaque = false
-        locationLabel.font = UIFont.systemFont(ofSize: 14, weight: .medium)
-        locationLabel.textColor = .gray
-        addSubview(locationLabel)
-        locationLabel.snp.makeConstraints { make in
-            make.leading.equalToSuperview().offset(16)
-            make.top.equalTo(statusLabel.snp.bottom).offset(8)
-            make.bottom.equalToSuperview().inset(16)
+        moreHoursButton.setImage(UIImage(named: "upArrow")?.withRenderingMode(.alwaysTemplate), for: .normal)
```
Comment (danielVebman) [inline Eatery/Views/Eateries/CampusMenuInfoView.swift:70]: I don't think you need to say `.withRenderingMode(.alwaysTemplate)`. Instead, you can set it to `Render As [Template Image]` in the icon's file in the assets folder.

File: Eatery/Views/Eateries/HoursThisWeekView.swift [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    private let diningMeals = ["Breakfast", "Lunch", "Dinner"]
+
+    init() {
+        super.init(frame: .zero)
+
+        segmentedControl.addTarget(self,
+                                   action: #selector(segmentedControlValueChanged(_:)),
+                                   for: .valueChanged)
+        addSubview(segmentedControl)
```
Comment (danielVebman) [inline Eatery/Views/Eateries/HoursThisWeekView.swift:34]: Is this how we do line wrapping? Pretty sure the Google style guide [says](https://google.github.io/swift/#function-calls) to do it like this:
```
segmentedControl.addTarget(
    self,
    action: #selector(...),
    for: .valueChanged
)
```
Also, [Tactile](https://github.com/delba/Tactile) provides a really nice condensed syntax (and support for closures!) if you're interested.

File: Eatery/Views/Eateries/HoursThisWeekView.swift [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        switch eatery.eateryType {
+        case .dining:
+            setShowDiningHall(true)
+
+            segmentedControl.removeAllSegments()
+            for meal in diningMeals {
+                segmentedControl.insertSegment(withTitle: meal,
+                                               at: segmentedControl.numberOfSegments,
+                                               animated: false)
+            }
```
Comment (danielVebman) [inline Eatery/Views/Eateries/HoursThisWeekView.swift:98]: Same thing about line wrapping.

File: Eatery/Views/Eateries/HoursThisWeekView.swift [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        } else {
+            for constraint in showDiningHallConstraints {
+                constraint.deactivate()
+            }
+            for constraint in hideDiningHallConstraints {
+                constraint.activate()
+            }
+        }
+    }
```
Comment (danielVebman) [inline Eatery/Views/Eateries/HoursThisWeekView.swift:162]: You could condense this function into:
```
for constraint in hideDiningHallConstraints {
    show ? constraint.deactivate() : constraint.activate()
}
for constraint in showDiningHallConstraints {
    show ? constraint.activate() : constraint.deactivate()
}
```
Or in two lines:
```
hideDiningHallConstraints.forEach { show ? $0.deactivate() : $0.activate() }
showDiningHallConstraints.forEach { show ? $0.activate() : $0.deactivate() }
```

**Review by danielVebman — APPROVED** [review]
You've got a lot of neat tricks in here! Just a couple little comments.

File: Eatery/Views/Eateries/CampusMenuInfoView.swift (hunk 6 above)
Comment (whoiswillma) [reply Eatery/Views/Eateries/CampusMenuInfoView.swift:70]: Eatery uses the up arrow image elsewhere and I didn't want to influence that

**Review by usubzero — APPROVED** [review]
looks good to go!

**Pull request merged**